c_cxx/QNN_EP/mobilenetv2_classification/main.cpp

```diff
@@ -0,0 +1,209 @@
+
+  OrtAllocator* allocator;
+  CheckStatus(g_ort, g_ort->GetAllocatorWithDefaultOptions(&allocator));
+  size_t num_input_nodes;
+  CheckStatus(g_ort, g_ort->SessionGetInputCount(session, &num_input_nodes));
+
+  std::vector<const char*> input_node_names;
+  std::vector<std::vector<int64_t>> input_node_dims;
+  std::vector<ONNXTensorElementDataType> input_types;
+  std::vector<OrtValue*> input_tensors;
+
+  input_node_names.resize(num_input_nodes);
+  input_node_dims.resize(num_input_nodes);
+  input_types.resize(num_input_nodes);
+  input_tensors.resize(num_input_nodes);
+
+  for (size_t i = 0; i < num_input_nodes; i++) {
+    // Get input node names
+    char* input_name;
+    CheckStatus(g_ort, g_ort->SessionGetInputName(session, i, allocator, &input_name));
+    input_node_names[i] = input_name;
+
+    // Get input node types
+    OrtTypeInfo* typeinfo;
+    CheckStatus(g_ort, g_ort->SessionGetInputTypeInfo(session, i, &typeinfo));
+    const OrtTensorTypeAndShapeInfo* tensor_info;
+    CheckStatus(g_ort, g_ort->CastTypeInfoToTensorInfo(typeinfo, &tensor_info));
+    ONNXTensorElementDataType type;
+    CheckStatus(g_ort, g_ort->GetTensorElementType(tensor_info, &type));
+    input_types[i] = type;
+
+    // Get input shapes/dims
+    size_t num_dims;
+    CheckStatus(g_ort, g_ort->GetDimensionsCount(tensor_info, &num_dims));
+    input_node_dims[i].resize(num_dims);
+    CheckStatus(g_ort, g_ort->GetDimensions(tensor_info, input_node_dims[i].data(), num_dims));
+
+    size_t tensor_size;
+    CheckStatus(g_ort, g_ort->GetTensorShapeElementCount(tensor_info, &tensor_size));
+
+    if (typeinfo) g_ort->ReleaseTypeInfo(typeinfo);
+  }
+
+  size_t num_output_nodes;
+  std::vector<const char*> output_node_names;
+  std::vector<std::vector<int64_t>> output_node_dims;
+  std::vector<OrtValue*> output_tensors;
+  CheckStatus(g_ort, g_ort->SessionGetOutputCount(session, &num_output_nodes));
+  output_node_names.resize(num_output_nodes);
+  output_node_dims.resize(num_output_nodes);
+  output_tensors.resize(num_output_nodes);
+
+  for (size_t i = 0; i < num_output_nodes; i++) {
+    // Get output node names
+    char* output_name;
+    CheckStatus(g_ort, g_ort->SessionGetOutputName(session, i, allocator, &output_name));
+    output_node_names[i] = output_name;
+
+    OrtTypeInfo* typeinfo;
+    CheckStatus(g_ort, g_ort->SessionGetOutputTypeInfo(session, i, &typeinfo));
+    const OrtTensorTypeAndShapeInfo* tensor_info;
+    CheckStatus(g_ort, g_ort->CastTypeInfoToTensorInfo(typeinfo, &tensor_info));
+
+    // Get output shapes/dims
+    size_t num_dims;
+    CheckStatus(g_ort, g_ort->GetDimensionsCount(tensor_info, &num_dims));
+    output_node_dims[i].resize(num_dims);
+    CheckStatus(g_ort, g_ort->GetDimensions(tensor_info, (int64_t*)output_node_dims[i].data(), num_dims));
+
+    size_t tensor_size;
+    CheckStatus(g_ort, g_ort->GetTensorShapeElementCount(tensor_info, &tensor_size));
+
+    if (typeinfo) g_ort->ReleaseTypeInfo(typeinfo);
```

Comment:
Do we want to use the C++ API in onnxruntime_cxx_api.h? The C++ APIs would take care of correctly releasing all resources via RAII, automatically throw exceptions if API calls return error status, and may reduce the amount of code.

Reply:
If we want to get this out soon, then we can also update in a separate follow-up PR.

Reply:
I think it depend on the user application.

---
In reply to: [1203270696](https://github.com/microsoft/onnxruntime-inference-examples/pull/246#discussion_r1203270696) [](ancestors = 1203270696)